### ISSUE TITLE
Classes are compiled with Java 11

### ISSUE THREAD

**Issue by doerfli** [camunda-community-hub/camunda-bpm-process-test-coverage#127]
On the github page is states that the library _Works with all relevant Java versions: 1.8 and 1.11 - using JUnit 4.13.1 (4.11 does not work) or JUnit 5_ but when i include version 1.0.0 in a Java 8 project, i get the error

```
.../BaseProcessTest.java:29: error: cannot access org.camunda.bpm.extension.process_test_coverage.junit.rules.MinimalCoverageMatcher
import org.camunda.bpm.extension.process_test_coverage.junit.rules.MinimalCoverageMatcher;
                                                                  ^
  bad class file: [sni]/org.camunda.bpm.extension/camunda-bpm-process-test-coverage-junit4/1.0.0/a12941963e854c29ec524fb03d1788dc6d022152/camunda-bpm-process-test-coverage-junit4-1.0.0.jar(org/camunda/bpm/extension/process_test_coverage/junit/rules/MinimalCoverageMatcher.class)
    class file has wrong version 55.0, should be 52.0
    Please remove or make sure it appears in the correct subdirectory of the classpath.
```

i assume the culprit is this entry in the pom.xml https://github.com/camunda-community-hub/camunda-bpm-process-test-coverage/blob/master/pom.xml#L26

If Java 8 should still be supported, then this needs to be fixed.

**Comment by zambrovski** [comment]
You are right, we will fix it.

**Comment by doerfli** [comment]
Hi @zambrovski ! Is there a date set for version _1.1_ yet? It would be helpful for us to get this in soon as we got issues with transitive dependencies of the old versions but cannot update to java 11 yet on our project. Thanks.

**Comment by zambrovski** [comment]
Tomorrow ?!

**Comment by doerfli** [comment]
Thursday is the earliest i will find time to integrate it, it might even slip into the first week of 2022. No rush there, i was just asking as the releases were not so often and i didn't want to wait another year. ;-) So year, tomorrow if fine 👍 Thank you very much! 😄

**Comment by zambrovski** [comment]
Since Java 8 is out of support already, it is easier to drop the support from the docs.
I encourage you to upgrade your software to JDK 11 (the next JDK 17 LTS is already out there!)

**Comment by doerfli** [comment]
unfortunately java 11 is currently not an option for our software (erp behemoth). we'll have to stick with the old one then. btw camunda still also supports java 8.

**Comment by zambrovski** [comment]
Let me try to just change the compiler level...

**Comment by doerfli** [comment]
Looks like build went through. I'll try to give it a spin in the next days. Thanks

**Comment by schonbachler** [comment]
Problem still exists with Version 1.0.2

**Comment by ingorichtsmeier** [comment]
Maybe it's a problem of the github workflows? 

They use still Java 11: https://github.com/camunda-community-hub/camunda-bpm-process-test-coverage/blob/master/.github/workflows/deploy.yml#L22

**Comment by ingorichtsmeier** [comment]
Hi @zambrovski,

as we use the test coverage in the java dev training from time to time training participants have only Java 8 available.

I've tried to build the software in a openjdk8 docker container and get this output from the surefire-report:

```
# cd surefire-reports
# ls
2022-02-02T08-31-46_569.dumpstream
# more 2022-02-02T08-31-46_569.dumpstream
# Created at 2022-02-02T08:31:47.183
Unrecognized option: --illegal-access=permit

# Created at 2022-02-02T08:31:47.185
Error: Could not create the Java Virtual Machine.

# Created at 2022-02-02T08:31:47.186
Error: A fatal exception has occurred. Program will exit.
```
Is it caused by the kotlin code used in the project? Or is it cauise by the build setup?

Is there any chance to get it complied with Java 8?

If not, I will add a note to the exercises to skip the test coverage, if you don't have Java 11 at least.

Thank you, Ingo

**Comment by zambrovski** [comment]
It has nothing to do with Kotlin but with the nice module system provided by Java and many libraries ignoring it.

**Comment by zambrovski** [comment]
Another question: the current version is compiled with JDK11 but having the JVM source and target version set to 1.8 (see https://github.com/camunda-community-hub/camunda-bpm-process-test-coverage/blob/a8c341e0dace27461451446b792cb16c3e264d7b/pom.xml#L27).

Do you say that the code is NOT working on JDK 1.8?

**Comment by ingorichtsmeier** [comment]
It's posted in the first comment of this thread: `class file has wrong version 55.0, should be 52.0`. 

I found some background information here: https://stackoverflow.com/a/56064482

But there may be other reasons.

I also quickly changed the github action yesterday to give it a try: https://github.com/camunda-community-hub/camunda-bpm-process-test-coverage/commit/4f5fe2e0110990035132ef8f149ad62d43c74637, but it failed.

tl;dr: I tried to compile with JDK 8 but it didn't work and throw an obscure error message which may be caused by the maven setup.

**Comment by ingorichtsmeier** [comment]
> I thought there is no reason to compile software using an outdated JDK.

The world is not always perfect, especially for companies with a signifcant amount of legacy systems...

**Comment by zambrovski** [comment]
Hello Ingo,

Let me sort this issue for you.

1. The code base is written in Kotlin and the compiler is not using anything above Java specification 1.8

2. As the issue was created we were using Java Source and Java Target level = 11. There is no difference what version of compiler you are using for it (everything above 11 is ok) and it created JDK11 classes (class version 55)

3. I FIXED this, by setting both values of Java Source and Java Target to 1.8 which should produces the Java classes compatible to JDK 8 (class version  52) In this case we could use any JDK >= 8 for compilation.

So my questions again:

1. Please post me the exact error you observe with JDK8. I don't have an example here in JDK 8 - but you have.

2. If a more or less decent JDK11 and Kotlin Compiler both support the options to produce JDK8 compatible classes, I prefer to use this, instead of run a compilation on a out-of-support JDK8 compiler. I do want to support JDK8 legacy customers, but  don't want to hack it somehow... The backport of the codebase and usage of JDK8 is possible, but is the last option I would take.

Be so kind and try out the example, if you want to help me fixing this issue, or just take some time and wait, until I'll find time to setup a non-working JDK8 example and can observe it on my own.

Cheers,

Simon

**Comment by zambrovski** [comment]
I believe I found the mistake. Will resolve it ASAP..

**Comment by ingorichtsmeier** [comment]
Hi @zambrovski,

Thank you for the insight. I believe, the most easy way to get a Java 8 compatible version of this exrension is to use Java 8 for compilation. And with the build by github actions (thanks again for setting this up), you can choose your build environment easily. But it's complicated to debug, if you don't see the root cause of the errors in the logs from the build runs.

I don't have Java 8 installed on my computer as well. So I pulled a docker image with jdk 8 and tried different maven builds.

1. `./mvnw clean package` fails when testing `core`
2. `./mvnw clean package -DskipTests=true` fails compiling report generator, as the Java version is set to 11 in the pom.
3. changing the Java version to 1.8, it fails compiling [CoverageReportUtils.java](https://github.com/camunda-community-hub/camunda-bpm-process-test-coverage/blob/3756778c439a80b07f4531afa6c6f3550faba5ea/extension/report-generator/src/main/java/org/camunda/bpm/extension/process_test_coverage/util/CoverageReportUtil.java#L50) as `Path.of` is a Java 11 statement.

Did you came to the same conclusion?
